client/browser/src/browser-extension/scripts/optionsPage.main.tsx

```diff
@@ -22,46 +19,94 @@ import {
+import { KnownCodeHost, knownCodeHosts } from '../knownCodeHosts'
+import { Optional } from 'utility-types'
+
+interface TabStatus {
+    host: string
+    protocol: string
+    hasPermissions: boolean
+    isPrivateRepository: boolean
+}
 
 assertEnvironment('OPTIONS')
 
 initSentry('options')
 
 const IS_EXTENSION = true
 
-interface State {
-    sourcegraphURL: string | null
-    isActivated: boolean
-    optionFlags: OptionFlagWithValue[]
-}
+/**
+ * A list of protocols where we should *not* show the permissions notification.
+ */
+const PERMISSIONS_PROTOCOL_BLOCKLIST = new Set(['chrome:', 'about:'])
+
+setLinkComponent(AnchorLink)
 
 const isOptionFlagKey = (key: string): key is OptionFlagKey =>
     !!optionFlagDefinitions.find(definition => definition.key === key)
 
-const fetchCurrentTabStatus = async (): Promise<OptionsMenuProps['currentTabStatus']> => {
+const fetchCurrentTabStatus = async (): Promise<TabStatus> => {
     const tabs = await browser.tabs.query({ active: true, currentWindow: true })
     if (tabs.length > 1) {
         throw new Error('Querying for the currently active tab returned more than one result')
     }
-    const { url } = tabs[0]
+    const { url, id } = tabs[0]
     if (!url) {
         throw new Error('Currently active tab has no URL')
     }
+    if (!id) {
+        throw new Error('Currently active tab has no ID')
+    }
+    const isPrivateRepository = await background.checkPrivateRepository(id)
     const { host, protocol } = new URL(url)
     const hasPermissions = await browser.permissions.contains({
         origins: [`${protocol}//${host}/*`],
     })
-    return { host, protocol, hasPermissions }
+    return { isPrivateRepository, host, protocol, hasPermissions }
 }
 
 // Make GraphQL requests from background page
-function requestGraphQL<T, V = object>(options: { request: string; variables: V }): Observable<GraphQLResult<T>> {
+function requestGraphQL<T, V = object>(options: {
+    request: string
+    variables: V
+    sourcegraphURL?: string
+}): Observable<GraphQLResult<T>> {
     return from(background.requestGraphQL<T, V>(options))
 }
 
+const version = getExtensionVersion()
+const isFullPage = !new URLSearchParams(window.location.search).get('popup')
+
+const validateSourcegraphUrl = (url: string): Observable<string | undefined> =>
+    fetchSite(options => requestGraphQL({ ...options, sourcegraphURL: url })).pipe(
+        mapTo(undefined),
+        catchError(error => {
+            const { message } = asError(error)
+            // We lose Error type when communicating from the background page
+            // to the options page, so we determine the error type from the message
+            if (message.includes('Failed to fetch')) {
+                return ['Incorrect Sourcegraph instance address']
+            }
+            if (message.includes('401')) {
+                return ['You are not authenticated']
```

Comment:
Ideally we would have something actionable here (a link to click for signing in), which we had before:

https://github.com/sourcegraph/sourcegraph/blob/963dc0f16934ba2deb746e9ab3b063f81181c018/client/browser/src/browser-extension/options-page/ServerUrlForm.tsx#L130-L140

This raises the question of whether we should support non-plain-text validation messages (`ReactNode`) so we can make them interactive.
Alternatively, we may wanna change the validation hook to only emit reason "error codes", so we can then have full expressive power in the render function to render them.

```suggestion
                return ['You are not signed in']
```